internal/usecase/interactor/user.go

```diff
@@ -241,6 +169,109 @@ func (i *User) Signup(ctx context.Context, inp interfaces.SignupParam) (u *user.
 	return u, team, nil
 }
 
+func (i *User) authSystemSignup(ctx context.Context, inp interfaces.SignupParam, tx repo.Tx) (repo.Tx, string, string, *user.User, *user.Team, error) {
+	if *inp.Name == "" {
+		return nil, "", "", nil, nil, interfaces.ErrSignupInvalidName
+	}
+	if _, err := mail.ParseAddress(*inp.Email); err != nil {
+		return nil, "", "", nil, nil, interfaces.ErrInvalidUserEmail
+	}
+	if *inp.Password == "" {
+		return nil, "", "", nil, nil, interfaces.ErrSignupInvalidPassword
+	}
+
+	tx, err := i.transaction.Begin()
+	if err != nil {
+		return nil, "", "", nil, nil, err
+	}
+	defer func() {
+		if err2 := tx.End(ctx); err == nil && err2 != nil {
+			err = err2
+		}
+	}()
+
+	// Check if user email already exists
+	existed, err := i.userRepo.FindByEmail(ctx, *inp.Email)
+	if err != nil && !errors.Is(err, rerror.ErrNotFound) {
+		return nil, "", "", nil, nil, err
+	}
+
+	if existed != nil {
+		if existed.Verification().IsVerified() {
+			return nil, "", "", nil, nil, errors.New("existed user email")
+		} else {
+			//	if user exists but not verified -> create a new verification
+			if err := i.CreateVerification(ctx, *inp.Email); err != nil {
+				return nil, "", "", nil, nil, err
+			} else {
+				team, err := i.teamRepo.FindByID(ctx, existed.Team())
+				if err != nil && !errors.Is(err, rerror.ErrNotFound) {
+					return nil, "", "", nil, nil, err
+				}
+				return nil, "", "", existed, team, nil
+			}
+		}
+	}
+
+	return tx, *inp.Name, *inp.Email, nil, nil, nil
+}
+
+func (i *User) auth0Signup(ctx context.Context, inp interfaces.SignupParam, tx repo.Tx) (repo.Tx, string, string, *user.Auth, error) {
+	if i.signupSecret != "" && *inp.Secret != i.signupSecret {
+		return nil, "", "", nil, interfaces.ErrSignupInvalidSecret
+	}
```

Comment:
Move also this validation out of the function to move codes to start transaction.